packages/puppeteer-core/src/common/Operation.ts

```diff
@@ -0,0 +1,131 @@
+/* eslint-disable tsdoc/syntax */
+
+import type {Awaitable} from './types.js';
+
+/**
+ * Operations are promises that can have
+ * [effects](<https://en.wikipedia.org/wiki/Side_effect_(computer_science)>)
+ * added on them (through {@link Operation.effect}).
+ *
+ * Semantically-speaking, adding an effect equates to guaranteeing the operation
+ * causes the added effect.
+ *
+ * The first effect that errors will propogate its error back to the operation.
+ *
+ * @example
+ *
+ * ```ts
+ * await input.click().effect(async () => {
+ *   await page.waitForNavigation();
+ * });
+ * ```
+ *
+ * @remarks
+ *
+ * Adding effects to a completed operation will result in an error. This occurs
+ * when either
+ *
+ * 1. the effects are added asynchronously or
+ * 2. the operation was awaited before effects were added.
+ *
+ * For example for (1),
+ *
+ * ```ts
+ * const operation = input.click();
+ * await new Promise(resolve => setTimeout(resolve, 100));
+ * await operation.effect(() => console.log('Works!')); // This will throw because of (1).
+ * ```
+ *
+ * For example for (2),
+ *
+ * ```ts
+ * const operation = await input.click();
+ * await operation.effect(() => console.log('Works!')); // This will throw because of (2).
+ * ```
+ *
+ * Tl;dr, effects **must** be added synchronously (no `await` statements between
+ * the time the operation is created and the effect is added).
+ *
+ * @public
+ */
+export class Operation<T> extends Promise<T> {
+  /**
+   * @internal
+   */
+  static create<T>(fn: () => Awaitable<T>, delay = 0): Operation<T> {
+    return new Operation((resolve, reject) => {
+      setTimeout(async () => {
+        try {
+          resolve(await fn());
+        } catch (error) {
+          reject(error);
+        }
+      }, delay);
+    });
+  }
+
+  #completed = false;
+  #effects: Array<Awaitable<unknown>> = [];
+  #error?: unknown;
+
+  /**
+   * Adds the given effect.
+   *
+   * @example
+   *
+   * ```ts
+   * await input.click().effect(async () => {
+   *   await page.waitForNavigation();
+   * });
+   * ```
+   *
+   * @param effect - The effect to add.
+   * @returns `this` for chaining.
+   */
+  effect(effect: () => Awaitable<void>): this {
+    if (this.#completed) {
+      throw new Error('Cannot add effect to a completed operation.');
+    }
+    try {
+      this.#effects.push(effect());
+    } catch (error) {
```

Comment:
why don't we throw in a sync way when effect is called? async errors won't be caught by this try/catch.